test/router.js

```diff
@@ -983,8 +985,117 @@ describe('Router', function () {
       .expect(200, 'saw GET /bar', done)
     })
   })
+
+  describe('events', function () {
+
+    describe('"handlestart"', function () {
+      it('should pass the request object', function (done) {
+        var router = new Router()
+        var server = createServer(router)
+
+        router.on('handlestart', function (req) {
+          assert(IncomingMessage.prototype.isPrototypeOf(req))
+          done()
+        })
+
+        request(server).get('/').end()
+      })
+
+      it('should be emitted at the beginning of handling the router stack', function (done) {
+        var router = new Router()
+        var server = createServer(router)
+
+        var handlers = [passThrough, passThrough, passThrough]
+        router.use(handlers)
+
+        var counter = 0
+        router.on('layer', function (layer, req) {
+          counter++
+        })
+        router.on('handlestart', function (req) {
+          assert.equal(0, counter)
+          done()
+        })
+
+        request(server).get('/').end()
+      })
+    })
+
+    describe('"layer"', function () {
+      it('should pass the matched layer and the request object', function (done) {
+        var router = new Router()
+        var server = createServer(router)
+
+        router.use(passThrough)
+
+        router.on('layer', function (layer, req) {
+          assert(Layer.prototype.isPrototypeOf(layer))
+          assert(IncomingMessage.prototype.isPrototypeOf(req))
+          done()
+        })
+
+        request(server).get('/').end()
+      })
+
+      it('should be emitted for each layer of the router stack', function (done) {
+        var router = new Router()
+        var server = createServer(router)
+
+        var handlers = [passThrough, passThrough, passThrough]
+        router.use(handlers)
+
+        var counter = 0
+        router.on('layer', function (layer, req) {
+          counter++
+          if (handlers.length === counter) done()
```

Comment:
I'd recommend using the `after` function for this type of tracking, which will throw when invoked too much.